README
Samuel 0.1.9
John Cheetham February 2019

Project Homepage : http://www.johncheetham.com/projects/samuel

Description
-----------
Samuel is a Draughts program written in python3/gtk3/C++.
It is licensed under the GPL (see the file named LICENSE).

It is aimed mainly at Linux users but should work on other systems
too if you have python3, gtk3, C++.

It is derived from guicheckers version 1.05. guicheckers is a draughts
program for windows which comes with source code. The C++ engine code,
board images and opening book/endgame databases all come from guicheckers.
So credit and thanks to the guicheckers people (programming by Jon Kreuzer
graphics by Josh Hess) for making their code available. You can find the
guicheckers website at http://www.3dkingdoms.com/checkers.htm. 

Samuel is named after Arthur Samuel, a pioneer of Computer Draughts, see
http://en.wikipedia.org/wiki/Arthur_Samuel.

Requirements
------------
Python3, gtk3, C++.


Installation
------------

To build it you will need python3, python-devel, gcc and gcc-c++
packages installed.

You can install these on debian 9:
    apt-get install build-essential python3-dev


Running Samuel from the Source Directory

    You can run samuel from the source directory without
    doing the full install.

    To do this on debian 9 for example:

        Enter 'python3.5 setup.py build' to build it.

        Then 'python3.5 run.py' to start samuel


Installing on the system     

    Enter 'python3.5 setup.py build' to build it (as normal user).
    Then 'python3.5 setup.py install' to install it (as root user).

    Note:

        Samuel should now be installed on your system. You can launch it from
        the gnome menu (under games) or type 'samuel' in any terminal window.

        There is no uninstall (distutils doesn't have one). If you need to     
        uninstall you have to make a note of the file names and then delete
        them manually.

        If running the build/install multiple times it's best to delete the
        build folder each time. 

Usage
-----
By default the computer plays white, you play red. In English Draughts Red
moves first. Click on a red piece and then click on a destination square to
move it. The computer will then automatically move a white piece.

If you want to play as white instead of red then select 'options' then
'computer plays red' from the menu bar. You will then need to click 'Go'
or press 'g' to start the game and make the computer make the first move
for red.

If you play as white you can also enable the 'Flip the Board' feature so
that the white pieces are at the bottom of the board.

If the computer plays both white and red you need to click the Go button
each time to make the computer move. If you are playing Human vs Computer
then the computer will make its move automatically after the Human move.

Use the menu bar to set the level of difficulty or save the game etc.
Use the buttons at the bottom to rewind the game.

If you want to set the lowest level then select user-defined level and
set the search depth to zero.

User settings are stored in the file ~/.samuel/settings
You can delete this file to go back to the default settings. 
You may also want to delete this file if you get problems when switching from
one version of samuel to another since the format can change between versions.

See also the online help file at
http://www.johncheetham.com/projects/samuel/help/help.html
This can be viewed in samuel from the help menu.

Summary of keyboard commands
----------------------------

    Keys        Function

    CTRL+N      New Game
    CTRL+O      Load Game
    CTRL+S      Save Game
    CTRL+Q      Quit Game
    m           Move Now - press this to make the computer move immediately
                when it's taking a long time thinking
    CTRL+C      Copy game to clipboard (in PDN format)
    CTRL+V      Paste game from clipbaord (in PDN format)
    CTRL+0      Reset board size to default size
    g           Same as the 'Go' button.
                Its main use is to make the computer make its move when the
                game is stopped.
    delete      Clear the board when in position edit mode 
    r           Retract last move
    [           rewind 1 move
    ]           forward 1 move
    {           rewind to start of game
    }           forward to end of game
    '3' - Used for opening book. See below.
    '2' - Used for opening book. See below.
    '4' - Used for opening book. See below.
    '6' - Used for opening book. See below.
    'K' - Used for opening book. See below.
    'S' - Used for opening book. See below.


Opening Book
------------
The file opening.gbk contains moves for the opening book.
It comes from guicheckers pre-seeded with opening moves.

You can modify it with these commands: 
    '3' - Add Current Board position / adjust towards '0'
    '2' - Add/Adjust to being good for red
    '4' - Add/Adjust to being good for white
    '6' - Remove Position
    'K' - Clear Opening Book in memory
    'S' - Save Opening Book

When saving the opening book is saved to ~/.samuel/opening.gbk.
When loading at startup it's loaded first from ~/.samuel/opening.gbk.
If not found there it will be loaded from the same directory as the program.
After a standard install the program directory will be read only which is why
'save' always saves to ~/.samuel/opening.gbk.

Most people will not want to modify the opening book. 


Troubleshooting
---------------
If it won't start after upgrading from an older version then try deleting the
file ~/.samuel/settings.

If you click on the buttons or press the keys repeatedly and very rapidly it
can break the engine. For this reason it's best to leave at least 1 second
between clicks.

End Game Database
-----------------
The files 2pc.cdb, 3pc.cdb and 4pc.cdb contain moves for the
end game. These files can be built using the genalldatabases
program.

Acknowledgements
----------------
Samuel uses C++ engine code, board graphics, opening book and endgame database
from guicheckers. 
guicheckers web page: http://www.3dkingdoms.com/checkers.htm



CHANGELOG
---------

Changes for 0.1.9
-----------------

2019-02-08  John Cheetham  developer@johncheetham.com

    * port to python3/gtk3

    * allow drag and drop of pieces

    * Make main window resizable

    * bugfixes
    
Changes for 0.1.8
-----------------

2009-10-27  John Cheetham  developer@johncheetham.com 

    * Add 'Flip the Board' feature

    * Allow player to choose colour to play (white or red).
      Also allow player vs computer, player vs player, computer vs computer

    * Add a status bar

    * Enforce time limit on user defined levels but not on pre-set levels.
      This will improve the play on the pre-set levels.

    * In position edit mode
         - Allow use of the Delete key to clear the board.
         - Ignore other key presses except for resize board

    * Add online help file

Changes for 0.1.7
-----------------

2009-10-02  John Cheetham  developer@johncheetham.com 

    * make board resizeable

    * remember users settings between program invocations

    * simplify level settings

    * fix bugs in gameover, movenow and time limit processing

    * add keyboard commands for retract move etc

    * clean up code


Changes for 0.1.6
-----------------

2009-09-05  John Cheetham  developer@johncheetham.com

    * nice display of multi jumps by the computer
    
    * edit board feature

    * suppress compiler warnings (though there are some on centos5 - must be an older compiler)

    * allow user-defined levels


Changes for 0.1.5
-----------------

2009-08-29  John Cheetham  developer@johncheetham.com

    * Display gameover at end of game

    * Fix keyboard shortcuts in menu

    * Fix fault after loading game from file/PDN/FEN when it's white to play
      Added a Go button to make white move

    * Don't hilight squares clicked on if end of game or white to move

    * Add explanatory messages to panel when using rewind keys.
        



debian/changelog
samuel (0.1) release; urgency=medium

  * Add locale de, es & it
  * Add dictionary  de.dict, en.dict, es.dict, fr.dict, it.dict, nl.dict and ru.dict
  * Add Stems it.stems and nl.stems 

 -- Thierry HUCHARD <thierry@ordissimo.com>  Sun, 11 Aug 2019 16:45:39 +0200




engine/movegen.cpp
/*
    movegen.cpp   
   
    This file is part of Samuel

    This file comes from the guicheckers project.
    See http://www.3dkingdoms.com/checkers.htm

    It has been amended for use in Samuel.

    Samuel is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    Samuel is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with Samuel.  If not, see <http://www.gnu.org/licenses/>.
   
 */

#define UINT unsigned int

#define WHITE 2
#define BLACK 1 // Black == Red
#define EMPTY 0
#define BPIECE 1
#define WPIECE 2
#define KING 4
#define BKING  5
#define WKING  6
#define INVALID 8
#define NONE 255
#define TIMEOUT 31000
#define DOUBLEJUMP 2000
#define HUMAN 128
#define MAKEMOVE 129
#define SEARCHED 127
#define MAX_SEARCHDEPTH 96
#define MAX_GAMEMOVES 2048
#define KING_VAL 33

// Convert between 32 square board and internal 44 square board
int BoardLocTo32[48] = { 0, 0, 0, 0, 0, 0, 1, 2, 3, 0, 4, 5, 6, 7, 8,  9, 10, 11, 0, 12, 13, 14, 15, 16, 17, 18, 19, 0, 20, 21, 22, 23, 24, 25, 26, 27, 0, 28, 29, 30, 31};
int BoardLoc32[32] = { 5, 6, 7, 8, 10, 11, 12, 13, 14, 15, 16, 17, 19, 20, 21, 22, 23, 24, 25, 26, 28, 29, 30, 31, 32, 33, 34, 35, 37, 38, 39, 40};

// =================================================
//
//              MOVE GENERATION 
//
// =================================================
struct SMove 
    {
     unsigned long SrcDst;
     char cPath[12];
    };

struct SCheckerBoard 
{
    unsigned long WP, BP, K;

    int GetBlackMoves( );
    int GetWhiteMoves( );
    UINT GetJumpersWhite( );
    UINT GetJumpersBlack( );
    UINT GetMoversWhite( );
    UINT GetMoversBlack( );
    void ConvertFromSqs( char sqs[] );
};

//
// BITBOARD INITILIZATION
//

//  28 29 30 31
// 24 25 26 27
//  20 21 22 23
// 16 17 18 19
//  12 13 14 15
// 08 09 10 11
//  04 05 06 07
// 00 01 02 03

UINT MASK_L3, MASK_L5, MASK_R3, MASK_R5, MASK_TOP, MASK_BOTTOM, MASK_TRAPPED_W, MASK_TRAPPED_B, MASK_2ND, MASK_7TH;
UINT MASK_EDGES, MASK_2CORNER1, MASK_2CORNER2;
UINT S[32];
unsigned char aBitCount[65536];
unsigned char aLowBit[65536];
unsigned char aHighBit[65536];

void InitBitTables()
{
    S[0] = 1;
    for (int i = 1; i < 32; i++) S[i] = S[i-1] * 2;

    MASK_L3 = S[1] | S[2] | S[3] | S[9] | S[10] | S[11] | S[17] | S[18] | S[19] | S[25] | S[26] | S[27];
    MASK_L5 = S[4] | S[5] | S[6] | S[12] | S[13] | S[14] | S[20] | S[21] | S[22];
    MASK_R3 = S[28] | S[29] | S[30] | S[20] | S[21] | S[22] | S[12] | S[13] | S[14] | S[4] | S[5] | S[6];
    MASK_R5 = S[25] | S[26] | S[27] | S[17] | S[18] | S[19] | S[9] | S[10] | S[11];
    MASK_TOP = S[28] | S[29] | S[30] | S[31] | S[24] | S[25] | S[26] | S[27] | S[20] | S[21] | S[22] | S[23];
    MASK_BOTTOM = S[0] | S[1] | S[2] | S[3] | S[4] | S[5] | S[6] | S[7] | S[8] | S[9] | S[10] | S[11];
    MASK_TRAPPED_W = S[0] | S[1] | S[2] | S[3] | S[4];
    MASK_TRAPPED_B = S[28] | S[29] | S[30] | S[31] | S[27];
    MASK_2ND = S[4] | S[5] | S[6] | S[7];
    MASK_7TH = S[24] | S[25] | S[26] | S[27];
    MASK_EDGES = S[24] | S[20] | S[16] | S[12] | S[8] | S[4] | S[0] | S[7] | S[11] | S[15] | S[19] | S[23] | S[27] | S[31];
    MASK_2CORNER1 = S[3] | S[7];
    MASK_2CORNER2 = S[24] | S[28];

    for (int Moves = 0; Moves < 65536; Moves++) {
        int nMoves = 0, nLow = 255, nHigh = 255;
        for (int i = 0; i < 16; i++) {
            if ( Moves & S[i] ) {
                nMoves++;
                if (nLow == 255) nLow = i;
                nHigh = i;
                }
        }
        aLowBit[ Moves ] = nLow;
        aHighBit[ Moves ] = nHigh;
        aBitCount[ Moves ] = nMoves;
    }
}

UINT inline BitCount( UINT Moves )
{
    if (Moves == 0) return 0;
    return aBitCount[ (Moves & 65535) ] + aBitCount[ ((Moves>>16) & 65535) ];
}

UINT inline FindLowBit( UINT Moves )
{
    if ( (Moves & 65535) ) return aLowBit[ (Moves & 65535) ];
    if ( ((Moves>>16) & 65535) ) return aLowBit[ ((Moves>>16) & 65535) ] + 16;
    return 0;
}

UINT inline FindHighBit( UINT Moves )
{
    if ( ((Moves>>16) & 65535) ) return aHighBit[ ((Moves>>16) & 65535) ] + 16;
    if ( (Moves & 65535) ) return aHighBit[ (Moves & 65535) ];
    return 0;
}

//
// BIT BOARD FUNCTIONS
//
UINT SCheckerBoard::GetMoversWhite( ) 
{
    const UINT nOcc = ~(WP | BP); // Not Occupied
    UINT Moves = (nOcc << 4);
    const UINT WK = WP&K;         // Kings
    Moves |= ((nOcc&MASK_L3) << 3);
    Moves |= ((nOcc&MASK_L5) << 5);
    Moves &= WP;
    if ( WK ) {
        Moves |= (nOcc >> 4) & WK;
        Moves |= ((nOcc&MASK_R3) >> 3) & WK;
        Moves |= ((nOcc&MASK_R5) >> 5) & WK;
        }    
    return Moves;
}

UINT SCheckerBoard::GetMoversBlack( ) 
{
    const UINT nOcc = ~(WP | BP);
    UINT Moves = (nOcc >> 4) & BP;
    const UINT BK = BP&K;
    Moves |= ((nOcc&MASK_R3) >> 3) & BP;
    Moves |= ((nOcc&MASK_R5) >> 5) & BP;
    if ( BK ) {
        Moves |= (nOcc << 4) & BK;
        Moves |= ((nOcc&MASK_L3) << 3) & BK;
        Moves |= ((nOcc&MASK_L5) << 5) & BK;
        }
    return Moves;
}

int SCheckerBoard::GetWhiteMoves( ) 
{
    const UINT nOcc = ~(WP | BP);   
    int nMoves = BitCount( (nOcc << 4) & WP );
    nMoves += BitCount( ( ((nOcc&MASK_L3) << 3) | ((nOcc&MASK_L5) << 5) ) & WP );
    const UINT WK = WP&K;   
    if ( WK ) {
        nMoves += BitCount( (nOcc >> 4) & WK );
        nMoves += BitCount( ( ((nOcc&MASK_R3) >> 3) | ((nOcc&MASK_R5) >> 5) ) & WK );
        }
    return nMoves;
}

int SCheckerBoard::GetBlackMoves( ) 
{
    const UINT nOcc = ~(WP | BP);
    int nMoves = BitCount( (nOcc >> 4) & BP );
    nMoves += BitCount( ( ((nOcc&MASK_R3) >> 3) | ((nOcc&MASK_R5) >> 5) ) & BP );
    const UINT BK = BP&K;
    if ( BK ) {
        nMoves += BitCount( (nOcc << 4) & BK );
        nMoves += BitCount( ( ((nOcc&MASK_L3) << 3) | ((nOcc&MASK_L5) << 5) ) & BK );
        }
    return nMoves;
}

UINT SCheckerBoard::GetJumpersWhite( )
{   
    const UINT nOcc = ~(WP | BP);
    const UINT WK = WP&K;
    UINT Movers = 0;
    UINT Temp = (nOcc << 4) & BP;
    if (Temp) Movers |= (((Temp&MASK_L3) << 3) | ((Temp&MASK_L5) << 5)) & WP;
    Temp = ( ((nOcc&MASK_L3) << 3) | ((nOcc&MASK_L5) << 5) ) & BP;
    Movers |= (Temp << 4) & WP;
    if (WK) {
        Temp = (nOcc>> 4) & BP;
        if (Temp) Movers |= (((Temp&MASK_R3) >> 3) | ((Temp&MASK_R5) >> 5)) & WK;
        Temp = ( ((nOcc&MASK_R3) >> 3) | ((nOcc&MASK_R5) >> 5) ) & BP;
        if (Temp) Movers |= (Temp >> 4) & WK;
        }   
    return Movers;
}

UINT SCheckerBoard::GetJumpersBlack( )
{   
    const UINT nOcc = ~(WP | BP);
    const UINT BK = BP&K;
    UINT Movers = 0;
    UINT Temp = (nOcc >> 4) & WP;
    if (Temp) Movers |= (((Temp&MASK_R3) >> 3) | ((Temp&MASK_R5) >> 5)) & BP;
    Temp = ( ((nOcc&MASK_R3) >> 3) | ((nOcc&MASK_R5) >> 5) ) & WP;
    Movers |= (Temp >> 4) & BP;
    if (BK) {
        Temp = (nOcc<< 4) & WP;
        if (Temp) Movers |= (((Temp&MASK_L3) << 3) | ((Temp&MASK_L5) << 5)) & BK;
        Temp = ( ((nOcc&MASK_L3) << 3) | ((nOcc&MASK_L5) << 5) ) & WP;
        if (Temp) Movers |= (Temp << 4) & BK;
        }
    return Movers;
}

void SCheckerBoard::ConvertFromSqs( char Sq[] )
{
    WP = 0;
    BP = 0;
    K = 0;
    for (int i = 0; i < 32; i++)
        {
        switch (Sq[ BoardLoc32[i] ]) {
            case WPIECE: WP |= S[i]; break;
            case BPIECE: BP |= S[i]; break;
            case WKING:  K |= S[i]; WP |= S[i]; break;
            case BKING:  K |= S[i]; BP |= S[i]; break;
            }
        }
}
//
//  E N D    B I T B O A R D S
//

struct CMoveList
    {
    // FUNCTIONS
    void inline Clear( ) {nJumps = 0;  
                          nMoves = 0;
                          }
    void inline SetMove( SMove &Move, int src, int dst)
        {
        Move.SrcDst = (src)+(dst<<6) + (1<<12);
        }
    void inline AddJump( SMove &Move, int pathNum)
        {
        Move.cPath[pathNum] = 0;
        Moves[ ++nJumps ] = Move;
        }
    void inline AddMove (int src, int dst)
        {
        Moves[ ++nMoves ].SrcDst = (src)+(dst<<6);
        return;
        }

    void inline AddSqDir( char board[], int square, int &nSqMoves, int nPiece, int pathNum, const int DIR, const int OPP_PIECE );
    void inline CheckJumpDir( char board[], int square, const int DIR, const int nOppPiece);
    void FindMovesBlack( char board[], SCheckerBoard &C);
    void FindMovesWhite( char board[], SCheckerBoard &C);
    void FindJumpsBlack( char board[], int Movers);
    void FindJumpsWhite( char board[], int Movers);
    void FindNonJumpsBlack( char board[], int Movers);
    void FindNonJumpsWhite( char board[], int Movers);
    void inline FindSqJumpsBlack( char board[], int square, int nPiece, int pathNum, int nJumpSq);
    void inline FindSqJumpsWhite( char board[], int square, int nPiece, int pathNum, int nJumpSq);

    // DATA
    int nMoves, nJumps;
    SMove Moves[36];
    SMove m_JumpMove;
    };

CMoveList g_Movelist[ MAX_SEARCHDEPTH + 1];
SMove g_GameMoves[ MAX_GAMEMOVES ];

// ------------------------------ FUNCTIONS --------------------------------
// Check for a normal move in DIR, then if a normal move is not possible, check for a jump
void inline CMoveList::CheckJumpDir( char board[], int square, const int DIR, const int nOppPiece )
{
    if ( (board[square + DIR ]&nOppPiece) && board[square + 2*DIR ] == EMPTY) 
        {// A jump is possible, call FindSqJumps to detect double/triple/etc. jumps
         SetMove (m_JumpMove, square, square + 2*DIR);
         if (nOppPiece == WPIECE) FindSqJumpsBlack( board, square + 2*DIR, board[ square ], 0, square + DIR);
         if (nOppPiece == BPIECE) FindSqJumpsWhite( board, square + 2*DIR, board[ square ], 0, square + DIR);
        } 
}
// -------------------------------------------------
// Find the Moves available on board, and store them in Movelist
// -------------------------------------------------
void CMoveList::FindMovesBlack( char board[], SCheckerBoard &C )
   {
    UINT Movers = C.GetJumpersBlack();
    if ( Movers ) FindJumpsBlack( board, Movers );
    else {
        FindNonJumpsBlack( board, C.GetMoversBlack() );
        }
  }
void CMoveList::FindMovesWhite( char board[], SCheckerBoard &C )
    {
    UINT Movers = C.GetJumpersWhite( );

    
    if ( Movers ) {       
        FindJumpsWhite( board, Movers );
    }
    else {
        FindNonJumpsWhite( board, C.GetMoversWhite( ) );
        }
    }

void CMoveList::FindNonJumpsWhite( char board[], int Movers )
    {    
    Clear( );
    while (Movers) 
        {
        UINT sq = FindLowBit( Movers );
        Movers &= ~S[sq];
        int square = BoardLoc32[ sq ];

        if ( board[square-5 ] == EMPTY ) AddMove( square, square-5); 
        if ( board[square-4 ] == EMPTY ) AddMove( square, square-4); 

        if (board[ square ] == WKING  ) {
            if ( board[square+4 ] == EMPTY ) AddMove( square, square+4); 
            if ( board[square+5 ] == EMPTY ) AddMove( square, square+5); 
            }   
        }
    }

void CMoveList::FindNonJumpsBlack( char board[], int Movers )
   {
    Clear( );
    while (Movers) 
        {
        UINT sq = FindHighBit( Movers );
        Movers &= ~S[sq];
        int square = BoardLoc32[ sq ];

        if ( board[square+4 ] == EMPTY ) AddMove( square, square+4); 
        if ( board[square+5 ] == EMPTY ) AddMove( square, square+5); 

        if (board[ square ] == BKING  ) {
            if ( board[square-5 ] == EMPTY ) AddMove( square, square-5); 
            if ( board[square-4 ] == EMPTY ) AddMove( square, square-4); 
            }
        }
  }

// -------------------------------------------------
// These two functions only add jumps
// -------------------------------------------------
void CMoveList::FindJumpsBlack( char board[], int Movers )
   {
    Clear( );
    while (Movers) 
        {
        UINT sq = FindHighBit( Movers );
        Movers &= ~S[sq];
        int square = BoardLoc32[ sq ];

        CheckJumpDir( board, square, 4, WPIECE );
        CheckJumpDir( board, square, 5, WPIECE );

        if (board[ square ] == BKING  ) {
            CheckJumpDir( board, square, -5, WPIECE );
            CheckJumpDir( board, square, -4, WPIECE );
            }
        }
    nMoves = nJumps;
  }

void CMoveList::FindJumpsWhite( char board[], int Movers )
    {
    Clear( );
    while (Movers) 
        {
        UINT sq = FindLowBit( Movers );
        Movers &= ~ S[sq];
        int square = BoardLoc32[ sq ];

        CheckJumpDir( board, square, -5, BPIECE );
        CheckJumpDir( board, square, -4, BPIECE );

        if (board[ square ] == WKING  ) {
             CheckJumpDir( board, square, 4, BPIECE );
             CheckJumpDir( board, square, 5, BPIECE );
            }
        }
    nMoves = nJumps;
    }

// -------------
//  If a jump move was possible, we must make the jump then check to see if the same piece can jump again.
//  There might be multiple paths a piece can take on a double jump, these functions store each possible path as a move.
// -------------
void inline CMoveList::AddSqDir( char board[], int square, int &nSqMoves, int nPiece, int pathNum, const int DIR, const int OPP_PIECE )
    {
    if ( (board[square + DIR ]&OPP_PIECE) && board[square + 2*DIR ] == EMPTY ) 
        {m_JumpMove.cPath[ pathNum ] = square + 2*DIR; 
         nSqMoves++; 
         if (OPP_PIECE == WPIECE) FindSqJumpsBlack( board, square+2*DIR, nPiece, pathNum+1, square + DIR);
         if (OPP_PIECE == BPIECE) FindSqJumpsWhite( board, square+2*DIR, nPiece, pathNum+1, square + DIR);
        }
    }

void CMoveList::FindSqJumpsBlack( char board[], int square, int nPiece, int pathNum, int nJumpSq)
{
    int nSqMoves = 0;
    // Remove the jumped piece (until the end of this function), so we can't jump it again
    int nJumpPiece = board[ nJumpSq ];
    board[ nJumpSq ] = EMPTY;

    // Now see if a piece on this square has more jumps
    AddSqDir ( board, square, nSqMoves, nPiece, pathNum, 4, WPIECE );
    AddSqDir ( board, square, nSqMoves, nPiece, pathNum, 5, WPIECE );
    if ( nPiece == BKING) 
        {// If this piece is a king, it can also jump backwards 
        AddSqDir ( board, square, nSqMoves, nPiece, pathNum, -5, WPIECE );
        AddSqDir ( board, square, nSqMoves, nPiece, pathNum, -4, WPIECE );
        }
    if (nSqMoves == 0) AddJump( m_JumpMove, pathNum); // this is a leaf, so store the move
    // Put back the jumped piece
    board[ nJumpSq ] = nJumpPiece;
}

void CMoveList::FindSqJumpsWhite( char board[], int square, int nPiece, int pathNum, int nJumpSq)
{
    int nSqMoves = 0;
    int nJumpPiece = board[ nJumpSq ];
    board[ nJumpSq ] = EMPTY;

    AddSqDir ( board, square, nSqMoves, nPiece, pathNum, -4, BPIECE );
    AddSqDir ( board, square, nSqMoves, nPiece, pathNum, -5, BPIECE );  
    if ( nPiece == WKING) 
        {
        AddSqDir ( board, square, nSqMoves, nPiece, pathNum, 5, BPIECE );
        AddSqDir ( board, square, nSqMoves, nPiece, pathNum, 4, BPIECE );   
        }
    if (nSqMoves == 0) AddJump( m_JumpMove, pathNum );
    board[ nJumpSq ] = nJumpPiece;
}
